File: Builder_Pattern/Typescript/Car.ts
export class Car {
  private carType: string;
  private seats: number;
  private engine: string;
  private transmission: string;
  private tripComputer: string;
  private gpsNavigator: string;
  private fuel: number = 0;

  public static getBuilder(): Builder {
    return new Builder();
  }

  private constructor(builder: Builder) {
    this.carType = builder.getCarType();
    this.seats = builder.getSeats();
    this.engine = builder.getEngine();
    this.transmission = builder.getTransmission();
    this.tripComputer = builder.getTripComputer();
    this.gpsNavigator = builder.getGpsNavigator();
  }

  public getCarType(): string {
    return this.carType;
  }

  public static createCarFromBuilder(builder: Builder): Car {
    return new Car(builder);
  }
}

class Builder {
  private carType: string;
  private seats: number;
  private engine: string;
  private transmission: string;
  private tripComputer: string;
  private gpsNavigator: string;

  public setCarType(carType: string): Builder {
    this.carType = carType;
    return this;
  }

  public getCarType(): string {
    return this.carType;
  }

  public setSeats(seats: number): Builder {
    this.seats = seats;
    return this;
  }

  public getSeats(): number {
    return this.seats;
  }

  public setEngine(engine: string): Builder {
    this.engine = engine;
    return this;
  }

  public getEngine(): string {
    return this.engine;
  }

  public setTransmission(transmission: string): Builder {
    this.transmission = transmission;
    return this;
  }

  public getTransmission(): string {
    return this.transmission;
  }

  public setTripComputer(tripComputer: string): Builder {
    this.tripComputer = tripComputer;
    return this;
  }

  public getTripComputer(): string {
    return this.tripComputer;
  }

  public setGpsNavigator(gpsNavigator: string): Builder {
    this.gpsNavigator = gpsNavigator;
    return this;
  }

  public getGpsNavigator(): string {
    return this.gpsNavigator;
  }

  public build(): Car {
    return Car.createCarFromBuilder(this);
  }
}

File: Builder_Pattern/Java/Car.java
public class Car{
    String carType;
    int seats;
    String engine;
    String transmission;
    String tripComputer;
    String gpsNavigator;
    double fuel = 0;

    public static Builder getBuilder(){
        return new Builder();
    }

    private Car(Builder builder){
        this.carType = builder.getCarType();
        this.seats = builder.getSeats();
        this.transmission = builder.getTransmission();
        this.tripComputer = builder.getTripComputer();
        this.gpsNavigator = builder.getGpsNavigator();
        this.fuel = builder.getFuel();
    }

    public String getCarType(){
        return carType;
    }


    public static class Builder{

        private String carType;
        private int seats;
        private String engine;
        private String transmission;
        private String tripComputer;
        private String gpsNavigator;
        private double fuel = 0;

        public String getCarType(){
            return carType;
        }

        public Builder setCarType(String carType){
            this.carType = carType;
            return this;
        }

        public int getSeats(){
            return seats;
        }

        public Builder setSeats(int seats){
            this.seats = seats;
            return this;
        }

        public String getRngine(){
            return engine;
        }

        public Builder setEngine(String engine){
            this.engine = engine;
            return this;
        }

        public String getTransmission(){
            return transmission;
        }

        public Builder setTransmission(String transmission){
            this.transmission = transmission;
            return this;
        }

        public String getTripComputer(){
            return tripComputer;
        }

        public Builder setTripComputer(String tripComputer){
            this.tripComputer = tripComputer;
            return this;
        }

        public String getGpsNavigator(){
            return gpsNavigator;
        }

        public Builder setGpsNavigator(String gpsNavigator){
            this.gpsNavigator = gpsNavigator;
            return this;
        }

        public double getFuel(){
            return fuel;
        }

        public Builder setFuel(double fuel){
            this.fuel = fuel;
            return this;
        }


        public Car build(){
            //validate the values
            if(getFuel() > 100){
                throw new IllegalArgumentException("Fuel limit is 100");
            }

            return new Car(this);
        }
        
    }




}